packages/transaction-messages/src/blockhash.ts

```diff
@@ -0,0 +1,71 @@
+import { SOLANA_ERROR__TRANSACTION__EXPECTED_BLOCKHASH_LIFETIME, SolanaError } from '@solana/errors';
+import { assertIsBlockhash, type Blockhash } from '@solana/rpc-types';
+
+import { IDurableNonceTransactionMessage } from './durable-nonce';
+import { BaseTransactionMessage } from './transaction-message';
+
+type BlockhashLifetimeConstraint = Readonly<{
+    blockhash: Blockhash;
+    lastValidBlockHeight: bigint;
+}>;
+
+export interface ITransactionMessageWithBlockhashLifetime {
+    readonly lifetimeConstraint: BlockhashLifetimeConstraint;
+}
+
+function isTransactionMessageWithBlockhashLifetime(
+    transaction: BaseTransactionMessage | (BaseTransactionMessage & ITransactionMessageWithBlockhashLifetime),
+): transaction is BaseTransactionMessage & ITransactionMessageWithBlockhashLifetime {
+    const lifetimeConstraintShapeMatches =
+        'lifetimeConstraint' in transaction &&
+        typeof transaction.lifetimeConstraint.blockhash === 'string' &&
+        typeof transaction.lifetimeConstraint.lastValidBlockHeight === 'bigint';
+    if (!lifetimeConstraintShapeMatches) return false;
+    try {
+        assertIsBlockhash(transaction.lifetimeConstraint.blockhash);
+        return true;
+    } catch {
+        return false;
+    }
+}
+
+export function assertIsTransactionMessageWithBlockhashLifetime(
+    transaction: BaseTransactionMessage | (BaseTransactionMessage & ITransactionMessageWithBlockhashLifetime),
+): asserts transaction is BaseTransactionMessage & ITransactionMessageWithBlockhashLifetime {
+    if (!isTransactionMessageWithBlockhashLifetime(transaction)) {
+        throw new SolanaError(SOLANA_ERROR__TRANSACTION__EXPECTED_BLOCKHASH_LIFETIME);
+    }
+}
+
+export function setTransactionMessageLifetimeUsingBlockhash<
```

Comment:
People are just absolutely going to hate me for this.

The method name really does explain what's _actually_ happening though.

Reply:
I don't think it's too bad tbh. 🙂